Add `-Zfunction-return={keep,thunk-extern}` option

ojeda opened the pull request:
This is intended to be used for Linux kernel RETHUNK builds.

With this commit (optionally backported to Rust 1.73.0), plus a patched Linux kernel to pass the flag, I get a RETHUNK build with Rust enabled that is `objtool`-warning-free and is able to boot in QEMU and load a sample Rust kernel module.

Issue: https://github.com/rust-lang/rust/issues/116853.

ojeda committed:
Add `-Zfunction-return={keep,thunk-extern}` option

This is intended to be used for Linux kernel RETHUNK builds.

With this commit (optionally backported to Rust 1.73.0), plus a
patched Linux kernel to pass the flag, I get a RETHUNK build with
Rust enabled that is `objtool`-warning-free and is able to boot in
QEMU and load a sample Rust kernel module.

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6562905111/job/17825927933) [(plain)](https://github.com/rust-lang/rust/commit/3cf45d69d2033e2c811c55acd5bb95ac821725e9/checks/17825927933/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ojeda
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_bcf8c599-6c3f-4064-9084-7d33a9db22c4
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=027977cd2e4781a98eac8147e31b012ad81a10c5
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_bcf8c599-6c3f-4064-9084-7d33a9db22c4
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_bcf8c599-6c3f-4064-9084-7d33a9db22c4
GITHUB_TRIGGERING_ACTOR=ojeda
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116892/merge
GITHUB_WORKFLOW_SHA=027977cd2e4781a98eac8147e31b012ad81a10c5
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
    Checking rustc_driver_impl v0.0.0 (/checkout/compiler/rustc_driver_impl)
error[E0433]: failed to resolve: use of undeclared type `FunctionReturn`
   --> compiler/rustc_interface/src/tests.rs:785:31
    |
785 |     tracked!(function_return, FunctionReturn::default());
    |                               ^^^^^^^^^^^^^^ use of undeclared type `FunctionReturn`
help: consider importing this enum
    |
2   + use rustc_session::config::FunctionReturn;
    |
```

</details>

ojeda commented:
Note that LLVM ignores the attribute in non-x86, but I have not replicated Clang's check that this only works in x86 for the CLI flag for the moment (should it be part of the target spec?).

Also, I linked the feature request issue as the tracking one in the docs -- we can reuse it changing the OP message, I assume, but please let me know if it is best to create a new one.

Nilstrieb commented on compiler/rustc_interface/src/tests.rs:
```diff
@@ -782,6 +783,7 @@ fn test_unstable_options_tracking_hash() {
     tracked!(flatten_format_args, false);
     tracked!(force_unstable_if_unmarked, true);
     tracked!(fuel, Some(("abc".to_string(), 99)));
+    tracked!(function_return, FunctionReturn::default());
```
this is supposed to use a non default value

ojeda replied:
Good catch, done.

Nilstrieb commented on compiler/rustc_session/src/config.rs:
```diff
@@ -3449,3 +3450,19 @@ impl Polonius {
+/// The different settings that the `-Zfunction-return` flag can have.
+#[derive(Clone, Copy, PartialEq, Hash, Debug)]
+pub enum FunctionReturn {
+    /// Keep the function return unmodified.
```
use `#[default]` and derive it

ojeda replied:
+1, done (we should change the rest in this file, by the way -- I was following `Polonius` which was recently added).

ojeda commented on compiler/rustc_codegen_llvm/src/attributes.rs:
```diff
@@ -333,6 +342,7 @@ pub fn from_fn_attrs<'ll, 'tcx>(
 
     // FIXME: none of these three functions interact with source level attributes.
     to_add.extend(frame_pointer_type_attr(cx));
+    to_add.extend(function_return_attr(cx));
     to_add.extend(instrument_function_attr(cx));
     to_add.extend(nojumptables_attr(cx));
     to_add.extend(probestack_attr(cx));
```
Yeah, it has been bad for a while (it was already there when I added `nojumptables`), but should probably be another PR (perhaps just remove "three"?).

ojeda replied:
Done in an extra commit.

bjorn3 commented on src/doc/unstable-book/src/compiler-flags/function-return.md:
```diff
@@ -0,0 +1,25 @@
+
+Like in Clang, GCC's values `thunk` and `thunk-inline` are not supported.
+
+Only x86 is supported.
```
Maybe add an error somewhere in rustc when trying to use it outside of x86?

ojeda replied:
Yeah, it would be probably be best to do so to avoid mistakes -- see https://github.com/rust-lang/rust/pull/116892#issuecomment-1768729744. GCC and Clang both also check that one is not using the `large` code model, so I will also add that.

ojeda replied:
I was talking to @nbdd0121 and he told me that a simple check is probably enough, i.e. no need to add state to the target spec if all x86 targets are supposed to support it.

ojeda replied:
Done, please take a look.

ojeda committed:
rustc_codegen_llvm: remove outdated count in comment

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

ojeda commented on compiler/rustc_session/src/session.rs:
```diff
@@ -1695,6 +1695,25 @@ fn validate_commandline_args_with_session_available(sess: &Session) {
+    match sess.opts.unstable_opts.function_return {
+        FunctionReturn::Keep => (),
+        FunctionReturn::ThunkExtern => {
+            // FIXME: In principle, the inherited base LLVM target code model could be large,
+            // but this only checks whether we were passed one explicitly (like Clang does).
```
If there is a way to get the effective code model, it would be nice to use it.

Urgau replied:
Maybe check [`sess.target.options.code_model`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/spec/struct.TargetOptions.html#structfield.code_model) ?

ojeda replied:
That covers even less, no? i.e. `code_model()` covers codegen and target options.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6580914013/job/17879820701) [(plain)](https://github.com/rust-lang/rust/commit/8b6016a6967f1e794eb33790cbf8fdcb4d4abdb5/checks/17879820701/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ojeda
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_7027e430-b4c2-4e57-9219-fbf83c749cfa
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=cf6fa6c8ac63a8143993890a78fbed94a034c87a
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_7027e430-b4c2-4e57-9219-fbf83c749cfa
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_7027e430-b4c2-4e57-9219-fbf83c749cfa
GITHUB_TRIGGERING_ACTOR=ojeda
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116892/merge
GITHUB_WORKFLOW_SHA=cf6fa6c8ac63a8143993890a78fbed94a034c87a
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container 6cb9003929cb
 ---> 2c25f72fb17d
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 01c95e19c0f9
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180117 sha256=2196c9034bf565528bbb1ee6dad4f753eb813f58822363e6b768f09c73e4d4ff
  Stored in directory: /tmp/pip-ephem-wheel-cache-4mzlfq9j/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.5-py3-none-any.whl (3.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.7/3.7 MB 18.9 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.12.4-py3-none-any.whl (11 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 38.9 MB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.12.4 platformdirs-3.11.0 virtualenv-20.24.5
Removing intermediate container ecfddff8bcb1
 ---> 638b88895f73
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 0f6032670aa8
 ---> 0f6032670aa8
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 3262309d7e23
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container f09e1d49640f
 ---> 0fb9a8a05bb2
Successfully built 0fb9a8a05bb2
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:0fb9a8a05bb2370b086327f970e60864b88d22e0f775581f77f4e0d76d2db9d2
Uploading finished image sha256:0fb9a8a05bb2370b086327f970e60864b88d22e0f775581f77f4e0d76d2db9d2 to https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
IMAGE          CREATED              CREATED BY                                      SIZE      COMMENT
0fb9a8a05bb2   1 second ago         /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
0f6032670aa8   2 seconds ago        /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
638b88895f73   3 seconds ago        |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
01c95e19c0f9   11 seconds ago       /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
2c25f72fb17d   12 seconds ago       |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      2 weeks ago          /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      2 weeks ago          /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      2 weeks ago          /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f6d4690f290>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
sha256:0fb9a8a05bb2370b086327f970e60864b88d22e0f775581f77f4e0d76d2db9d2
---
DirectMap4k:      257984 kB
DirectMap2M:     7081984 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/94c4e5c4111cbefcd098d6efcb0b222b6e512fcb/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-94c4e5c4111cbefcd098d6efcb0b222b6e512fcb-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished release [optimized] target(s) in 26.23s
##[endgroup]
fmt check
tidy check
tidy error: following path contains more than 865 entries, you should move the test to some relevant subdirectory (current: 866): /checkout/tests/ui
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (23.2.1)
Collecting pip
  Obtaining dependency information for pip from https://files.pythonhosted.org/packages/e0/63/b428aaca15fcd98c39b07ca7149e24bc14205ad0f1c80ba2b01835aedde1/pip-23.3-py3-none-any.whl.metadata
  Downloading pip-23.3-py3-none-any.whl.metadata (3.5 kB)
Downloading pip-23.3-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 23.2.1
    Uninstalling pip-23.2.1:
    Uninstalling pip-23.2.1:
      Successfully uninstalled pip-23.2.1
Successfully installed pip-23.3
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 13.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 13.7 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 86.2 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
WARNING: There was an error checking the latest version of pip.
some tidy checks failed
Build completed unsuccessfully in 0:00:50
  local time: Thu Oct 19 22:31:16 UTC 2023
  network time: Thu, 19 Oct 2023 22:31:16 GMT
```

</details>

ojeda commented:
I added a folder but that still counts against the tidy limit in `tests/ui`. Should I move the folder to e.g. `tests/ui/invalid-compile-flags` or increase the limit by 1?

ojeda commented:
One more note: GCC and Clang give an error if one tries to use `function-return` at all, i.e. even with `keep`, on non-x86. However, in this PR `keep` is allowed for all architectures.

I did it like this because it seemed simpler (no `Option`) and more flexible, especially since (in the case of this flag) other architectures could support particular modes, e.g. s390 expolines (https://github.com/llvm/llvm-project/blob/49af6502c6dcb4a7f7520178bd14df396f78240c/clang/lib/Frontend/CompilerInvocation.cpp#L1942).

But perhaps consistency with GCC/Clang is better?

bors commented:
:umbrella: The latest upstream changes (presumably #117405) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ojeda commented:
Resolved trivial formatting conflict.

wesleywiser commented:
Thanks for already getting the tracking issue set up @ojeda!

@bors r+

bors commented:
:pushpin: Commit e4f42de5d6193b60c0b752cd2e722e10e31a3751 has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser e4f42de5d6193b60c0b752cd2e722e10e31a3751 -->
<!-- homu: {"type":"Approved","sha":"e4f42de5d6193b60c0b752cd2e722e10e31a3751","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e4f42de5d6193b60c0b752cd2e722e10e31a3751 with merge 7e6b53374d592fae909caf1b4e4cf5b93c5ba19f...
<!-- homu: {"type":"BuildStarted","head_sha":"e4f42de5d6193b60c0b752cd2e722e10e31a3751","merge_sha":"7e6b53374d592fae909caf1b4e4cf5b93c5ba19f"} -->

rust-log-analyzer commented:
The job **`x86_64-apple-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6726762327/job/18283571563) [(plain)](https://github.com/rust-lang-ci/rust/commit/7e6b53374d592fae909caf1b4e4cf5b93c5ba19f/checks/18283571563/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
SCCACHE_BUCKET=rust-lang-ci-sccache2
SCRIPT=./x.py --stage 2 test --skip tests/ui --skip tests/rustdoc --skip tests/run-make-fulldeps
SHELL=/bin/bash
SHLVL=3
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.Chrgk5dVbA/Listeners
STATS_EXT=true
STATS_EXTP=https://provjobdsettingscdn.blob.core.windows.net/settings/provjobdsettings-0.5.154/provjobd.data
STATS_RDCL=true
STATS_TIS=mining
---
failures:

---- [assembly] tests/assembly/x86_64-function-return.rs#keep-thunk-extern stdout ----

error in revision `keep-thunk-extern`: verification with 'FileCheck' failed
status: exit status: 1
command: "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/ci-llvm/bin/FileCheck" "--input-file" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/assembly/x86_64-function-return.keep-thunk-extern/x86_64-function-return.s" "/Users/runner/work/rust/rust/tests/assembly/x86_64-function-return.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC,keep-thunk-extern" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/Users/runner/work/rust/rust/tests/assembly/x86_64-function-return.rs:24:24: error: keep-thunk-extern: expected string not found in input
 // keep-thunk-extern: jmp __x86_return_thunk
                       ^
/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/assembly/x86_64-function-return.keep-thunk-extern/x86_64-function-return.s:5:6: note: scanning from here
_foo:
     ^
/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/assembly/x86_64-function-return.keep-thunk-extern/x86_64-function-return.s:13:2: note: possible intended match here
 jmp ___x86_return_thunk


Input file: /Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/assembly/x86_64-function-return.keep-thunk-extern/x86_64-function-return.s


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1:  .section __TEXT,__text,regular,pure_instructions 
            2:  .macosx_version_min 10, 12 
            3:  .globl _foo 
            4:  .p2align 4, 0x90 
            5: _foo: 
check:24'0          X error: no match found
            6:  .cfi_startproc 
check:24'0     ~~~~~~~~~~~~~~~~
            7:  pushq %rbp 
check:24'0     ~~~~~~~~~~~~
            8:  .cfi_def_cfa_offset 16 
check:24'0     ~~~~~~~~~~~~~~~~~~~~~~~~
            9:  .cfi_offset %rbp, -16 
check:24'0     ~~~~~~~~~~~~~~~~~~~~~~~
           10:  movq %rsp, %rbp 
check:24'0     ~~~~~~~~~~~~~~~~~
           11:  .cfi_def_cfa_register %rbp 
check:24'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           12:  popq %rbp 
check:24'0     ~~~~~~~~~~~
           13:  jmp ___x86_return_thunk 
check:24'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
check:24'1      ?                        possible intended match
           14:  .cfi_endproc 
check:24'0     ~~~~~~~~~~~~~~
           15:  
check:24'0     ~
           16: .subsections_via_symbols 
check:24'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------


---- [assembly] tests/assembly/x86_64-function-return.rs#thunk-extern stdout ----
---- [assembly] tests/assembly/x86_64-function-return.rs#thunk-extern stdout ----

error in revision `thunk-extern`: verification with 'FileCheck' failed
status: exit status: 1
command: "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/ci-llvm/bin/FileCheck" "--input-file" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/assembly/x86_64-function-return.thunk-extern/x86_64-function-return.s" "/Users/runner/work/rust/rust/tests/assembly/x86_64-function-return.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC,thunk-extern" "--dump-input-context" "100"
--- stderr -------------------------------
/Users/runner/work/rust/rust/tests/assembly/x86_64-function-return.rs:22:19: error: thunk-extern: expected string not found in input
/Users/runner/work/rust/rust/tests/assembly/x86_64-function-return.rs:22:19: error: thunk-extern: expected string not found in input
 // thunk-extern: jmp __x86_return_thunk
                  ^
/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/assembly/x86_64-function-return.thunk-extern/x86_64-function-return.s:5:6: note: scanning from here
_foo:
     ^
/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/assembly/x86_64-function-return.thunk-extern/x86_64-function-return.s:13:2: note: possible intended match here
 jmp ___x86_return_thunk


Input file: /Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/assembly/x86_64-function-return.thunk-extern/x86_64-function-return.s


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1:  .section __TEXT,__text,regular,pure_instructions 
            2:  .macosx_version_min 10, 12 
            3:  .globl _foo 
            4:  .p2align 4, 0x90 
            5: _foo: 
check:22'0          X error: no match found
            6:  .cfi_startproc 
check:22'0     ~~~~~~~~~~~~~~~~
            7:  pushq %rbp 
check:22'0     ~~~~~~~~~~~~
            8:  .cfi_def_cfa_offset 16 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~
            9:  .cfi_offset %rbp, -16 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~
           10:  movq %rsp, %rbp 
check:22'0     ~~~~~~~~~~~~~~~~~
           11:  .cfi_def_cfa_register %rbp 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           12:  popq %rbp 
check:22'0     ~~~~~~~~~~~
           13:  jmp ___x86_return_thunk 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
check:22'1      ?                        possible intended match
           14:  .cfi_endproc 
check:22'0     ~~~~~~~~~~~~~~
           15:  
check:22'0     ~
           16: .subsections_via_symbols 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6726762327/job/18284528010)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6726762327/job/18284528010","builder_name":"checks-actions"} -->

raoulstrackx commented on tests/assembly/x86_64-function-return.rs:
```diff
@@ -0,0 +1,28 @@
+// CHECK-LABEL: foo:
+#[no_mangle]
+pub unsafe fn foo() {
+    // unset: ret
```
In order to mitigate LVI vulnerabilities, the `x86_64-fortanix-unknown-sgx` target must not use `ret` instructons. Every `ret` is rewritten into something like `popq %rax; lfence; jmp *%rax`. Therefore these tests fail on that platform. Can this platform be ignored for these tests please?

ojeda replied:
Added `// ignore-sgx Tests incompatible with LVI mitigations` like you did in other tests.

raoulstrackx commented on tests/assembly/x86_64-function-return.rs:
```diff
@@ -0,0 +1,28 @@
+    // thunk-extern-NOT: ret
+    // keep-thunk-extern: jmp __x86_return_thunk
+    // keep-thunk-extern-NOT: ret
+    // thunk-extern-keep: ret
```
Same issue as previous

raoulstrackx reviewed (commented):
This PR breaks CI for the `x86_64-fortanix-unknown-sgx` platform. Can this be fixed before it's merged please?

cc: @vn971 @jethrogb

bors commented:
:umbrella: The latest upstream changes (presumably #117706) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@ojeda any updates on this? thanks

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7046120167/job/19177262606) [(plain)](https://github.com/rust-lang/rust/commit/06e4fb747ce3c5047dedb8dd2c640fd768809b0a/checks/19177262606/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ojeda
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_94f9c9e4-3701-4675-b94e-7146973c635d
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=05cf53317e96bf761cffa9c1049730e666f57f9e
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_94f9c9e4-3701-4675-b94e-7146973c635d
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_94f9c9e4-3701-4675-b94e-7146973c635d
GITHUB_TRIGGERING_ACTOR=ojeda
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116892/merge
GITHUB_WORKFLOW_SHA=05cf53317e96bf761cffa9c1049730e666f57f9e
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container fb2a78953fe1
 ---> 951857903c43
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> ea5784e9fe55
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-lxzaac4p/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.7-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 68.4 MB/s eta 0:00:00
Collecting platformdirs<5,>=3.9.1
  Downloading platformdirs-4.0.0-py3-none-any.whl (17 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 124.1 MB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.13.1 platformdirs-4.0.0 virtualenv-20.24.7
Removing intermediate container a6be764cb713
 ---> d33eb607f058
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> c21ff4dd9bcd
 ---> c21ff4dd9bcd
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 3641548f147e
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container b5a06550bfb2
 ---> 96c932c0b012
Successfully built 96c932c0b012
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:96c932c0b012a217babccedfa7f390737a0737bba523ffb1a597350d6dd71987
Uploading finished image sha256:96c932c0b012a217babccedfa7f390737a0737bba523ffb1a597350d6dd71987 to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
96c932c0b012   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
c21ff4dd9bcd   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
d33eb607f058   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.8MB    
ea5784e9fe55   9 seconds ago    /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
951857903c43   11 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      8 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      8 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      8 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f773b30a4d0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:96c932c0b012a217babccedfa7f390737a0737bba523ffb1a597350d6dd71987
---
DirectMap4k:      184256 kB
DirectMap2M:     7155712 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/1670ff64bf1ccb2ad71068254b53725631c55864/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-1670ff64bf1ccb2ad71068254b53725631c55864-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 25.74s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_session/src/session.rs at line 1693:
         FunctionReturn::ThunkExtern => {
             // FIXME: In principle, the inherited base LLVM target code model could be large,
             // but this only checks whether we were passed one explicitly (like Clang does).
-            if let Some(code_model) = sess.code_model() && code_model == CodeModel::Large {
+            if let Some(code_model) = sess.code_model()
+                && code_model == CodeModel::Large
+            {
                 sess.emit_err(errors::FunctionReturnThunkExternRequiresNonLargeCodeModel);
         }
         }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_session/src/output.rs" "/checkout/compiler/rustc_session/src/filesearch.rs" "/checkout/compiler/rustc_session/src/cstore.rs" "/checkout/compiler/rustc_session/src/session.rs" "/checkout/compiler/rustc_session/src/code_stats.rs" "/checkout/compiler/rustc_session/src/config.rs" "/checkout/compiler/rustc_attr/src/session_diagnostics.rs" "/checkout/compiler/rustc_session/src/config/sigpipe.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Nov 30 12:05:01 UTC 2023
  network time: Thu, 30 Nov 2023 12:05:02 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

ojeda commented:
Rebased, added `// ignore-dawrin` and `// ignore-sgx`.

For Darwin, if there is a "standard" way to check for the extra underscore, please let me know (relaxing the regex felt wrong, but perhaps duplicating the cases is OK for this sort of test). Otherwise, LLVM's own test use `linux-gnu` for these tests, so I guess it is OK.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7046275910/job/19177751302) [(plain)](https://github.com/rust-lang/rust/commit/45c13873e4b442f72bd6d6e76588fc0c3bacc664/checks/19177751302/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ojeda
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_2e7a0c2d-ca62-407c-bf4e-9bb4088ac415
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=ab8d5bb0de91bdb549bc8219fc9734347d6ef937
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_2e7a0c2d-ca62-407c-bf4e-9bb4088ac415
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_2e7a0c2d-ca62-407c-bf4e-9bb4088ac415
GITHUB_TRIGGERING_ACTOR=ojeda
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116892/merge
GITHUB_WORKFLOW_SHA=ab8d5bb0de91bdb549bc8219fc9734347d6ef937
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
........................................................................................  6072/15876
........................................................................................  6160/15876
........................................................................................  6248/15876
.....................................................................i..............i.i.  6336/15876
...............iiiiii.............................................................F...F.  6424/15876
........................................................................................  6600/15876
........................................................................................  6688/15876
........................................................................................  6776/15876
........................................................................................  6864/15876
---

---- [ui] tests/ui/invalid-compile-flags/function-return/requires-x86-or-x86_64.rs#aarch64 stdout ----
diff of stderr:

1 error: `-Zfunction-return` (except `keep`) is only supported on x86 and x86_64
- error: aborting due to previous error
+ error: aborting due to 1 previous error
4 
5 
5 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid-compile-flags/function-return/requires-x86-or-x86_64.aarch64/requires-x86-or-x86_64.aarch64.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args invalid-compile-flags/function-return/requires-x86-or-x86_64.rs`

error in revision `aarch64`: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/invalid-compile-flags/function-return/requires-x86-or-x86_64.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "aarch64" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid-compile-flags/function-return/requires-x86-or-x86_64.aarch64" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid-compile-flags/function-return/requires-x86-or-x86_64.aarch64/auxiliary" "-Zfunction-return=thunk-extern" "--target" "aarch64-unknown-linux-gnu"
--- stderr -------------------------------
--- stderr -------------------------------
error: `-Zfunction-return` (except `keep`) is only supported on x86 and x86_64
error: aborting due to 1 previous error
------------------------------------------



---- [ui] tests/ui/invalid-compile-flags/function-return/thunk-extern-requires-non-large-code-model.rs#large stdout ----
diff of stderr:

1 error: `-Zfunction-return=thunk-extern` is only supported on non-large code models
- error: aborting due to previous error
Build completed unsuccessfully in 0:12:51
+ error: aborting due to 1 previous error
4 
4 
5 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid-compile-flags/function-return/thunk-extern-requires-non-large-code-model.large/thunk-extern-requires-non-large-code-model.large.stderr
To only update this specific test, also pass `--test-args invalid-compile-flags/function-return/thunk-extern-requires-non-large-code-model.rs`


error in revision `large`: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/invalid-compile-flags/function-return/thunk-extern-requires-non-large-code-model.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "large" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid-compile-flags/function-return/thunk-extern-requires-non-large-code-model.large" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid-compile-flags/function-return/thunk-extern-requires-non-large-code-model.large/auxiliary" "--target" "x86_64-unknown-linux-gnu" "-Zfunction-return=thunk-extern" "-Ccode-model=large"
--- stderr -------------------------------
--- stderr -------------------------------
error: `-Zfunction-return=thunk-extern` is only supported on non-large code models
error: aborting due to 1 previous error
------------------------------------------


```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7051198365/job/19193554867) [(plain)](https://github.com/rust-lang/rust/commit/4c86c7980100f2a548c0e69e1a6728a2272b808f/checks/19193554867/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ojeda
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_12707e7c-136e-46fb-81cb-b336f62a79c5
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=e7ff6ed132b8d1e008af424c2416bc7754d21272
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_12707e7c-136e-46fb-81cb-b336f62a79c5
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_12707e7c-136e-46fb-81cb-b336f62a79c5
GITHUB_TRIGGERING_ACTOR=ojeda
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116892/merge
GITHUB_WORKFLOW_SHA=e7ff6ed132b8d1e008af424c2416bc7754d21272
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container 0cbaacf2d46f
 ---> a23791ae2512
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 0a4a0bbd56c8
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-wbt5l86e/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.7-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 19.8 MB/s eta 0:00:00
Collecting platformdirs<5,>=3.9.1
  Downloading platformdirs-4.0.0-py3-none-any.whl (17 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.13.1 platformdirs-4.0.0 virtualenv-20.24.7
Removing intermediate container dd60b6771c27
 ---> 8eca7a3130ef
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> d416f620a030
 ---> d416f620a030
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> fa336d79c8f8
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 91e2d39c4699
 ---> e5911155bcb6
Successfully built e5911155bcb6
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:e5911155bcb63b12b78beed9a5890e31f8b232ce4fd98da699b8fde1d1c0676f
Uploading finished image sha256:e5911155bcb63b12b78beed9a5890e31f8b232ce4fd98da699b8fde1d1c0676f to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
e5911155bcb6   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
d416f620a030   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
8eca7a3130ef   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.8MB    
0a4a0bbd56c8   10 seconds ago   /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
a23791ae2512   12 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      8 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      8 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      8 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f213596e590>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:e5911155bcb63b12b78beed9a5890e31f8b232ce4fd98da699b8fde1d1c0676f
---
DirectMap4k:      198592 kB
DirectMap2M:     6092800 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/07921b50ba6dcb5b2984a1dba039a38d85bffba2/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-07921b50ba6dcb5b2984a1dba039a38d85bffba2-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/tests/assembly/x86_64-function-return.rs:12: line longer than 100 chars
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (23.3.1)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 12.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 12.4 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 68.5 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:00:59
  local time: Thu Nov 30 19:19:01 UTC 2023
  network time: Thu, 30 Nov 2023 19:19:02 GMT
```

</details>

ojeda commented:
A bunch of trivial fixes above.

@wesleywiser @Dylan-DPC This is ready again -- let's see if other targets fail in the full CI.

wesleywiser commented:
Thanks for the rebase @ojeda!

@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 2d476222e8458aa873e6760aac189b5e9d0a9930 has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser 2d476222e8458aa873e6760aac189b5e9d0a9930 -->
<!-- homu: {"type":"Approved","sha":"2d476222e8458aa873e6760aac189b5e9d0a9930","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 2d476222e8458aa873e6760aac189b5e9d0a9930 with merge f45631b10f2199a7ef344a936af9eb60342c16ee...
<!-- homu: {"type":"BuildStarted","head_sha":"2d476222e8458aa873e6760aac189b5e9d0a9930","merge_sha":"f45631b10f2199a7ef344a936af9eb60342c16ee"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7052849711/job/19201472428)
Approved by: wesleywiser
Pushing f45631b10f2199a7ef344a936af9eb60342c16ee to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7052849711/job/19201472428"},"merge_sha":"f45631b10f2199a7ef344a936af9eb60342c16ee"} -->

rust-timer commented:
Finished benchmarking commit (f45631b10f2199a7ef344a936af9eb60342c16ee): [comparison URL](https://perf.rust-lang.org/compare.html?start=87e1447aadaa2899ff6ccabe1fa669eb50fb60a1&end=f45631b10f2199a7ef344a936af9eb60342c16ee&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=87e1447aadaa2899ff6ccabe1fa669eb50fb60a1&end=f45631b10f2199a7ef344a936af9eb60342c16ee&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=87e1447aadaa2899ff6ccabe1fa669eb50fb60a1&end=f45631b10f2199a7ef344a936af9eb60342c16ee&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-1.8%, -1.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-1.3%, -1.3%] | 1     |
| All ❌✅ (primary)                 | -1.8% | [-1.8%, -1.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=87e1447aadaa2899ff6ccabe1fa669eb50fb60a1&end=f45631b10f2199a7ef344a936af9eb60342c16ee&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean   | range           | count |
|:----------------------------------:|:------:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -      | -               | 0     |
| Regressions ❌ <br /> (secondary)  | -      | -               | 0     |
| Improvements ✅ <br /> (primary)   | -14.8% | [-21.4%, -9.2%] | 7     |
| Improvements ✅ <br /> (secondary) | -      | -               | 0     |
| All ❌✅ (primary)                 | -14.8% | [-21.4%, -9.2%] | 7     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=87e1447aadaa2899ff6ccabe1fa669eb50fb60a1&end=f45631b10f2199a7ef344a936af9eb60342c16ee&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 674.44s -> 674.191s (-0.04%)
**Artifact size**: 313.39 MiB -> 313.41 MiB (0.01%)

ojeda commented:
@wesleywiser Thanks for the very quick review and have a nice holiday :)

wesleywiser commented:
Thanks! ☺️
